evm/src/fuzz/invariant/error.rs

```diff
@@ -0,0 +1,224 @@
+            addr: invariant_contract.address,
+            func,
+            inner_sequence: inner_sequence.to_vec(),
+        }
+    }
+
+    /// Replays the error case and collects all necessary traces.
+    pub fn replay(
+        &self,
+        mut executor: Executor,
+        known_contracts: Option<&ContractsByArtifact>,
+        mut ided_contracts: ContractsByAddress,
+        logs: &mut Vec<Log>,
+        traces: &mut Vec<(TraceKind, CallTraceArena)>,
+    ) -> Option<CounterExample> {
+        let mut counterexample_sequence = vec![];
+        let calls = match self.test_error {
+            // Don't use at the moment.
+            TestError::Abort(_) => return None,
+            TestError::Fail(_, ref calls) => calls,
+        };
+
+        let calls = self.try_shrinking(calls, &executor);
+
+        // We want traces for a failed case.
+        executor.set_tracing(true);
+
+        set_up_inner_replay(&mut executor, &self.inner_sequence);
+
+        // Replay each call from the sequence until we break the invariant.
+        for (sender, (addr, bytes)) in calls.iter() {
+            let call_result = executor
+                .call_raw_committing(*sender, *addr, bytes.0.clone(), 0.into())
+                .expect("bad call to evm");
+
+            logs.extend(call_result.logs);
+            traces.push((TraceKind::Execution, call_result.traces.clone().unwrap()));
+
+            // Identify newly generated contracts, if they exist.
+            ided_contracts.extend(load_contracts(
+                vec![(TraceKind::Execution, call_result.traces.unwrap())],
+                known_contracts,
+            ));
+
+            counterexample_sequence.push(BaseCounterExample::create(
+                *sender,
+                *addr,
+                bytes,
+                &ided_contracts,
+            ));
+
+            // Checks the invariant.
+            if let Some(func) = &self.func {
+                let error_call_result = executor
+                    .call_raw(CALLER, self.addr, func.0.clone(), 0.into())
+                    .expect("bad call to evm");
+
+                if error_call_result.reverted {
+                    logs.extend(error_call_result.logs);
+                    traces.push((TraceKind::Execution, error_call_result.traces.unwrap()));
+                    break
+                }
+            }
+        }
+
+        (!counterexample_sequence.is_empty())
+            .then_some(CounterExample::Sequence(counterexample_sequence))
+    }
+
+    /// Tests that the sequence of calls successfully reverts on the error function.
+    fn fails_successfully(&self, mut executor: Executor, calls: &[&BasicTxDetails]) -> bool {
+        for (sender, (addr, bytes)) in calls {
+            executor
+                .call_raw_committing(*sender, *addr, bytes.0.clone(), 0.into())
+                .expect("bad call to evm");
+
+            // Checks the invariant.
+            if let Some(func) = &self.func {
```

Comment:
Nit - can pull this outside of the loop and `return false` early, before iterating?

Reply:
Slightly modified it, but the goal here is to check that the invariant has, in fact, been broken after a call. 

So pulling this outside of the loop is not what we want